File: android/src/main/kotlin/com/pdfbarcode/pdf_barcode_decoder/PdfDecodeManager.kt
package com.pdfbarcode.pdf_barcode_decoder

import android.content.Context
import android.graphics.pdf.PdfRenderer
import android.os.ParcelFileDescriptor
import com.pdfbarcode.pdf_barcode_decoder.renderer.AndroidPdfPageRenderer
import com.pdfbarcode.pdf_barcode_decoder.scanner.IouDeduplicator
import com.pdfbarcode.pdf_barcode_decoder.scanner.ZXingBarcodeScanner
import java.io.File
import java.io.FileOutputStream
import java.util.UUID

class PdfDecodeManager(private val context: Context) {

    fun decodePdf(
        pdfBytes: ByteArray?,
        filePath: String?,
        config: Map<String, Any?>
    ): List<Map<String, Any>> {
        val file = if (pdfBytes != null) {
            val tempFile = File(context.cacheDir, "pdf_bc_${UUID.randomUUID()}.pdf")
            FileOutputStream(tempFile).use { it.write(pdfBytes) }
            tempFile
        } else if (filePath != null) {
            File(filePath)
        } else {
            throw IllegalArgumentException("Either pdfBytes or filePath must be provided")
        }

        val renderer = AndroidPdfPageRenderer()
        val scanner = ZXingBarcodeScanner()
        val resultList = mutableListOf<Map<String, Any>>()

        try {
            val pfd = try {
                ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
            } catch (e: Exception) {
                throw Exception("INVALID_PDF: Could not open file")
            }
            
            val pdfRenderer = try {
                PdfRenderer(pfd)
            } catch (e: SecurityException) {
                throw Exception("ENCRYPTED_PDF: PDF is password protected")
            } catch (e: Exception) {
                throw Exception("INVALID_PDF: Invalid PDF format")
            }
            
            val pageCount = pdfRenderer.pageCount

            val dpi = (config["dpi"] as? Int) ?: 300
            val firstPageOnly = (config["firstPageOnly"] as? Boolean) ?: false
            val stopAfterFirst = (config["stopAfterFirst"] as? Boolean) ?: false
            val maxPages = (config["maxPages"] as? Int) ?: pageCount
            @Suppress("UNCHECKED_CAST")
            val formats = (config["formats"] as? List<String>) ?: listOf("all")

            val pagesToScan = if (firstPageOnly) 1 else minOf(pageCount, maxPages)

            for (i in 0 until pagesToScan) {
                val page = pdfRenderer.openPage(i)
                try {
                    val bitmap = renderer.renderPage(page, dpi)
                    val rawResults = scanner.scan(bitmap, formats)
                    bitmap.recycle()

                    val cleanResults = IouDeduplicator.deduplicate(rawResults)
                    for (barcode in cleanResults) {
                        resultList.add(
                            mapOf(
                                "type" to barcode.type,
                                "value" to barcode.value,
                                "page" to i,
                                "left" to barcode.left,
                                "top" to barcode.top,
                                "width" to barcode.width,
                                "height" to barcode.height
                            )
                        )
                    }

                    if (stopAfterFirst && resultList.isNotEmpty()) {
                        break
                    }
                } finally {
                    page.close()
                }
            }

            pdfRenderer.close()
            pfd.close()
        } finally {
            if (pdfBytes != null && file.exists()) {
                file.delete()
            }
        }

        return resultList
    }
}

File: ios/Classes/PdfDecodeManager.swift
import Foundation
import PDFKit

class PdfDecodeManager {
    func decodePdf(pdfBytes: FlutterStandardTypedData?, filePath: String?, config: [String: Any]) throws -> [[String: Any]] {
        let url: URL
        var isTemp = false
        
        if let pdfBytes = pdfBytes {
            let tempUrl = FileManager.default.temporaryDirectory.appendingPathComponent("pdf_bc_\(UUID().uuidString).pdf")
            try pdfBytes.data.write(to: tempUrl)
            url = tempUrl
            isTemp = true
        } else if let filePath = filePath {
            url = URL(fileURLWithPath: filePath)
            print("[PdfDecodeManager] Using file: \(url.lastPathComponent), exists: \(FileManager.default.fileExists(atPath: filePath))")
        } else {
            throw NSError(domain: "INVALID_ARGS", code: 1, userInfo: [NSLocalizedDescriptionKey: "Either pdfBytes or filePath must be provided"])
        }
        
        defer {
            if isTemp {
                try? FileManager.default.removeItem(at: url)
            }
        }
        
        print("[PdfDecodeManager] Opening PDF document")
        guard let document = PDFDocument(url: url) else {
            print("[PdfDecodeManager] ERROR: PDFDocument failed to open document")
            throw NSError(domain: "INVALID_PDF", code: 2, userInfo: [NSLocalizedDescriptionKey: "Could not open PDF file"])
        }
        print("[PdfDecodeManager] PDF opened: \(document.pageCount) page(s), encrypted=\(document.isEncrypted)")
        
        if document.isEncrypted {
            throw NSError(domain: "ENCRYPTED_PDF", code: 3, userInfo: [NSLocalizedDescriptionKey: "PDF is password protected"])
        }
        
        let renderer = DefaultPdfPageRenderer()
        let scanner = AppleVisionBarcodeScanner()
        var resultList = [[String: Any]]()
        
        let dpi = config["dpi"] as? Int ?? 300
        let firstPageOnly = config["firstPageOnly"] as? Bool ?? false
        let stopAfterFirst = config["stopAfterFirst"] as? Bool ?? false
        let maxPages = config["maxPages"] as? Int ?? document.pageCount
        let formats = config["formats"] as? [String] ?? ["all"]
        
        print("[PdfDecodeManager] Config: dpi=\(dpi), firstPageOnly=\(firstPageOnly), stopAfterFirst=\(stopAfterFirst), maxPages=\(maxPages), formats=\(formats)")
        
        let pagesToScan = firstPageOnly ? 1 : min(document.pageCount, maxPages)
        print("[PdfDecodeManager] Will scan \(pagesToScan) page(s)")
        
        for i in 0..<pagesToScan {
            guard let page = document.page(at: i) else {
                print("[PdfDecodeManager] WARNING: Could not get page at index \(i)")
                continue
            }
            
            print("[PdfDecodeManager] --- Processing page \(i) ---")
            if let image = renderer.renderPage(page, dpi: dpi) {
                let rawResults = scanner.scan(image: image, formats: formats)
                print("[PdfDecodeManager] Page \(i): found \(rawResults.count) barcode(s)")
                for var barcode in rawResults {
                    barcode["page"] = i
                    resultList.append(barcode)
                }
                
                if stopAfterFirst && !resultList.isEmpty {
                    break
                }
            } else {
                print("[PdfDecodeManager] WARNING: renderPage returned nil for page \(i)")
            }
        }
        
        print("[PdfDecodeManager] Total barcodes found: \(resultList.count)")
        return resultList
    }
}
